config_types.cpp

```diff
@@ -263,3 +266,93 @@ const char* server_addr::get_last_error(void) const
 {
     return gai_strerror(m_last_error);
 }
+
+config_cpu_list::config_cpu_list(unsigned int cores)
+{
+    for (unsigned int i = 0; i < cores; ++i)
+        cpu_list.push_back(i);
+
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list::config_cpu_list(std::string str)
+{
+    std::stringstream ss(str);
+
+    if (str.find(' ') != std::string::npos)
+        return;
+
+    char separator = str[str.find_first_of(",-")];
+
+    int i;
+    while (ss >> i)
+    {
+        cpu_list.push_back(i);
+
+        if (ss.peek() == separator)
+            ss.ignore();
+        else if (ss.eof())
+            break;
+        else {
+            cpu_list.clear();
+            break;
+        }
+    }
+
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list::config_cpu_list(const config_cpu_list& copy) :
+        cpu_list(copy.cpu_list)
+{
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list& config_cpu_list::operator=(const config_cpu_list& cl)
+{
+    if (this == &cl)
+        return *this;
+
+    cpu_list = cl.cpu_list;
+    next_cpu = cpu_list.begin();
+    return *this;
+}
+
+bool config_cpu_list::is_defined() const
+{
+    return cpu_list.size() > 0;
+}
+
+bool config_cpu_list::is_legal() const
+{
+    unsigned int max_core = get_nprocs();
+
+    for (std::vector<unsigned int>::const_iterator it = cpu_list.begin(); it != cpu_list.end(); ++it) {
```

Comment:
is there a `max()` function in cpp, that you can use instead of iterating the list by yourself?
for example - 
```cpp
unsigned int max_core = max<unsigned int>(cpu_list);
return max_core <= get_nproc();
```